### DISCUSSION TITLE
Google authentication

### DISCUSSION THREAD

**Discussion by petteri111** [zauberzeug/nicegui#385] [Q&A]
Hello, has anyone made a Google Oauth2 authentication to a nicegui webpage? ChatGPT gave me following:

> To add Google OAuth to a NiceGUI web page, you can follow these general steps:
> 
> 1. Create a Google Cloud Platform project and set up an OAuth client ID. You will need to set the "Authorized redirect URIs" to the URL of your NiceGUI web page. Take note of the "Client ID" and "Client Secret" values, as you will need them later.
> 
> 2. Install the google-auth and google-auth-oauthlib Python packages, which provide the necessary functionality for authenticating with Google OAuth.
> 
> 3. In your NiceGUI web page code, create a link that directs the user to Google's OAuth consent page. The link should include the following parameters:
> 
>     `client_id`: The client ID you obtained from Google.
>     `redirect_uri`: The URL of your NiceGUI web page that will handle the OAuth response.
>     `response_type`: Set this to "code" to indicate that you want to receive an authorization code that can be exchanged for an access token.
>     `scope`: The list of Google API scopes that your web page needs access to.
>     `access_type`: Set this to "offline" if you need a refresh token in addition to the access token.
> 
> 4. When the user clicks on the link, they will be directed to Google's OAuth consent page. Once the user grants access to your web page, they will be redirected to the redirect_uri that you specified. In your NiceGUI web page code, you need to handle the authorization code that is included in the redirect URL. You can use the google-auth-oauthlib package to exchange the authorization code for an access token and refresh token (if applicable).
> 
> 5. Once you have obtained an access token, you can use it to make authenticated requests to Google APIs from your NiceGUI web page. The google-auth package provides a simple way to create an authorized API client.
> 
> Here is some sample Python code that demonstrates the general flow:
> 
> ```python
> from google.oauth2.credentials import Credentials
> from google_auth_oauthlib.flow import InstalledAppFlow
> from googleapiclient.discovery import build
> 
> # Step 3: Create a link to Google's OAuth consent page
> client_id = '<your client ID>'
> redirect_uri = '<your redirect URI>'
> scope = ['https://www.googleapis.com/auth/calendar.readonly']
> access_type = 'offline'
> auth_url = f'https://accounts.google.com/o/oauth2/auth?client_id={client_id}&redirect_uri={redirect_uri}&response_type=code&scope={" ".join(scope)}&access_type={access_type}'
> 
> # Step 4: Handle the authorization code
> # Assuming the authorization code is in the 'code' URL parameter
> flow = InstalledAppFlow.from_client_config({'client_id': client_id, 'client_secret': '<your client secret>'}, scopes=scope)
> credentials = flow.fetch_token(code=<authorization code>)
> 
> # Step 5: Make an API request
> service = build('calendar', 'v3', credentials=credentials)
> events_result = service.events().list(calendarId='primary', timeMin='2023-02-15T00:00:00Z', maxResults=10, singleEvents=True, orderBy='startTime').execute()
> ```
> 
> Note that this is just a basic example and you may need to modify the code to fit your specific needs. Additionally, you should ensure that you are following Google's guidelines for OAuth, including properly securing your client secrets and using authorized redirect URIs.

**Comment by falkoschindler** [comment]
Interesting. When I ask ChatGPT if it knows a Python library called NiceGUI, it says it doesn't. And I think it doesn't really know about "NiceGUI websites" either. But the procedure looks rather generic. Did you try it?

**Comment by rodja** [comment]
I suggest you ask ChatGPT: "How to use google oauth2 with fastAPI?" because they do not have trained the model on data after 2021; it has no knowledge of NiceGUI, jet :-(
We specifically designed NiceGUI to run on top of FastAPI for being able to handle all kinds of backend related features. You may want to look at [our FastAPI app integration example](https://github.com/zauberzeug/nicegui/tree/main/examples/fastapi/) and [our auth demo](https://github.com/zauberzeug/nicegui/tree/main/examples/authentication) which adapts the [OAuth2 example from FastAPI](https://fastapi.tiangolo.com/tutorial/security/simple-oauth2/).

**Reply by phoskee** [reply]
It seems that ChatGPT is familiar with NiceGui, despite saying 

> I am trained using OpenAI's data up until September 2021.

It's likely that it has been updated, but he himself doesn't seem to know yet.
🤣

**Comment by rodja** [comment]
As a side note I suggest you do not copy & past ChatGPT answers directly. It kind of pollutes the internet.

**Comment by rodja** [comment]
I just stumbled upon https://github.com/yigitoo/auth-site from @yigitoo. Looks like a Google OAuth implementation. Have not tried but wanted to share the link here :-)

**Reply by o12486vs2** [reply]
I've been testing based on the code from [this repository](https://github.com/yigitoo/auth-site). However, I consistently encountered a KeyError stating that `'id_token'` couldn't be found when using this line: 
```python
user = await oauth.google.parse_id_token(request, token)
```

To address this, I referred to the [OAuth documentation](https://docs.authlib.org/en/latest/client/starlette.html). According to the documentation:

```python
token = await oauth.google.authorize_access_token(request)
user = token.get('userinfo')
if user:
    request.session['user'] = user
    userEmail = user['email']
```

It appears that the `token` already uses the `id_token` to retrieve the `userinfo`, which is then returned as a dictionary. Hence, we can directly use the `userinfo` from the `token`.
